pkg/ipcache/ipcache.go

```diff
@@ -536,8 +548,22 @@ func (ipc *IPCache) deleteLocked(ip string, source source.Source) (namedPortsCha
 		namedPortsChanged = ipc.updateNamedPorts()
 	}
 
-	if newHostIP != nil {
-		nodeID = NodeHandler.AllocateNodeID(newHostIP)
+	// If this is a straight delete this is simple, deallocate the nodeID. If its a
+	// Upsert then the deleted IP was shadowing another CIDR so we will restore
+	// that mapping. But, from NodeID side we already did an allocate for that and
+	// we don't want to bump the refcnt here so simply Get() the ID. Remember to
+	// delete the old IP through the Upsert path though.
+	if cacheModification == Delete {
+		NodeHandler.DeallocateNodeID(oldHostIP)
+	} else if cacheModification == Upsert {
```

Comment:
Incidentally we can probably be pretty confident that the upsert case here is when there is a CIDR prefix `/32` in policy that is shadowed by a real pod Identity, so if we delete the pod then restore the CIDR prefix (hence `cacheModification == Upsert`), that CIDR prefix has no host association so the "newHostIP" case should not be possible. Probably also the newHostIP is always non-nil. Furthermore, probably (though I say this with a little less confidence), the oldHostIP is always set.

It's not expressible in the current ipcache structure to have two competing pods with the same IP such that when you delete one of those pods, we put the other one (with a corresponding hostIP) into the ipcache.

All that said, the code here would theoretically do the right thing, I just think that a couple of the conditional statements here are likely never hit.

Reply:
I agree, but I was trying to be a bit defensive to handle whatever the ipcache threw at us.